crates/wasm/src/client/implementation/specific.rs

```diff
@@ -261,3 +272,147 @@ impl wit::client_window::Host for Bindings {
+        let texture = Texture::new_with_data(
+            gpu,
+            &wgpu::TextureDescriptor {
+                label: None,
+                size: wgpu::Extent3d {
+                    width: desc.width,
+                    height: desc.height,
+                    depth_or_array_layers: 1,
+                },
+                mip_level_count: 1,
+                sample_count: 1,
+                dimension: wgpu::TextureDimension::D2,
+                format: desc.format.from_bindgen(),
+                usage: wgpu::TextureUsages::TEXTURE_BINDING,
+                view_formats: &[],
+            },
+            &desc.data,
+        );
+        let texture = Arc::new(texture);
+        let texture_view = Arc::new(texture.create_view(&TextureViewDescriptor::default()));
+        let storage = world.resource_mut(procedural_storage());
+        let texture_handle = storage.insert_texture(texture_view);
+        Ok(texture_handle.into_bindgen())
+    }
+    fn destroy(&mut self, handle: wit::client_texture::Handle) -> anyhow::Result<()> {
+        let world = self.world_mut();
+        let storage = world.resource_mut(procedural_storage());
+        storage.remove_texture(handle.from_bindgen());
+        Ok(())
+    }
+}
+impl wit::client_sampler::Host for Bindings {
+    fn create(
+        &mut self,
+        desc: wit::client_sampler::Descriptor,
+    ) -> anyhow::Result<wit::client_sampler::Handle> {
+        let world = self.world_mut();
+        let assets = world.resource(asset_cache());
+        let gpu = GpuKey.get(assets);
+        let sampler = gpu.device.create_sampler(&wgpu::SamplerDescriptor {
+            address_mode_u: desc.address_mode_u.from_bindgen(),
+            address_mode_v: desc.address_mode_v.from_bindgen(),
+            address_mode_w: desc.address_mode_w.from_bindgen(),
+            mag_filter: desc.mag_filter.from_bindgen(),
+            min_filter: desc.min_filter.from_bindgen(),
+            mipmap_filter: desc.mipmap_filter.from_bindgen(),
+            ..wgpu::SamplerDescriptor::default()
+        });
+        let sampler = Arc::new(sampler);
+        let storage = world.resource_mut(procedural_storage());
+        let sampler_handle = storage.insert_sampler(sampler);
+        Ok(sampler_handle.into_bindgen())
+    }
+    fn destroy(&mut self, handle: wit::client_sampler::Handle) -> anyhow::Result<()> {
+        let world = self.world_mut();
+        let storage = world.resource_mut(procedural_storage());
+        storage.remove_sampler(handle.from_bindgen());
+        Ok(())
+    }
+}
+impl wit::client_material::Host for Bindings {
+    fn create(
+        &mut self,
+        desc: wit::client_material::Descriptor,
+    ) -> anyhow::Result<wit::client_material::Handle> {
+        let world = self.world_mut();
+        let storage = world.resource_mut(procedural_storage());
+        let material = PbrMaterialConfig {
+            source: "Procedural Material".to_string(),
```

Comment:
It might be a good idea to include a `source` in the WIT descriptor, so that users can figure out what material/mesh/etc they're dealing with when looking at a lot of procedural content in a dev tool that shows this